### ISSUE TITLE
SDL_AudioStreamAvailable incorrectly states that zero data is in the stream if less than the sample size is still present

### ISSUE THREAD

**Issue by leonstyhre** [libsdl-org/SDL#4947]
Hi!

Example scenario:

My application has 1024 set as the sample size/amount of samples.
Audio is playing, then for some reason playback is interrupted when there are 880 samples left to be processed.
Calling SDL_AudioStreamAvailable then returns 0. 
In this situation, if calling SDL_AudioStreamGet, the 880 remaining samples are played first, followed by any new samples that have been added to the stream. So there is clearly data in the stream, and SDL_AudioStreamGet can retrieve it even if SDL_AudioStreamAvailable says there is nothing there.

As well, if instead adding for example 10000 samples via SDL_AudioStreamPut and then calling SDL_AudioStreamAvailable the value returned is 10880, which is the correct value.

The issue seems to occur on all SDL versions I've tried and at least on BSD Unix and Linux and probably on Windows and macOS too.

Here's some example work-around code I use in my application:

    std::vector<Uint8> writeBuffer(10000);
    if (SDL_AudioStreamPut(sConversionStream, reinterpret_cast<const void*>(&writeBuffer.at(0)),
                           10000) == -1) {
        LOG(LogError) << "Failed to put samples in the conversion stream:";
        LOG(LogError) << SDL_GetError();
        mIsClearingStream = false;
        return;
    }

    int length = SDL_AudioStreamAvailable(sConversionStream);

    std::vector<Uint8> readBuffer(length);
    SDL_AudioStreamGet(sConversionStream, static_cast<void*>(&readBuffer.at(0)), length);

This code is necessary because SDL_AudioStreamClear is unstable and causes random crashes on BSD Unix, Linux, macOS and Windows and the only way I've found to safely empty the stream is to mute the audio, call SDL_AudioStreamGet and then unmuting again.

Thanks in advance!

**Comment by icculus** [comment]
> because SDL_AudioStreamClear is unstable and causes random crashes

Let's talk about that before we look further into this bug, because this sounds like a memory corruption bug that might be in SDL, or might be in your code.

It would be _extremely_ useful to see a small piece of code that reproduces this, or your application run under Valgrind to see it misbehaving somewhere near the SDL_AudioStreamClear call.


As to the original bug report: it's not unexpected for data to be put into an audio stream and SDL_AudioStreamAvailable() to report zero, as it might need a fairly large buffer to resample properly. If you keep adding data, it will eventually start reporting it available when it has enough queued up to safely resample the input without producing audio artifacts.

If you absolutely need the converted data before then, you can call SDL_AudioStreamFlush(), which will treat the input as complete and resample as if any further data it would need is silence...but you cannot flush an ongoing stream without introducing audio artifacts, as you're effectively padding the existing input data with silence in between each additional Put.

tl;dr: At the end of the stream, explicitly Flush to get the final data out of it. If the stream is continuing, just add more as it shows up and accept more output as it moves towards being available.

(If this isn't causing the problem you're seeing, we can dig in further to figure out if this is an SDL bug, but the "AudioStreamClear is unreliable" is a huge red flag for deeper problems and we _really_ should resolve that first.)

**Comment by icculus** [comment]
(Also, this should move to the 2.0.20 milestone unless we determine soon that this is really an SDL bug.)

**Comment by leonstyhre** [comment]
Thanks for your feedback!

The issue is that all these three functions have been unstable in my experience:
SDL_AudioStreamClear
SDL_AudioStreamFlush
SDL_FreeAudioStream

So I had to use this workaround of clearing out the stream manually to accomplish what I needed, which is that when a new video starts in my application, the user should not hear a short audio snippet from the last video played. I agree that I would normally not need to use this approach assuming the above functions would be completely stable.

I'll try to capture some more data for you to help pinpoint the issue, but it is quite rare, it happens perhaps every 200 times one of these functions is called but maybe more often during some (for me) unknown circumstances. But that's enough to make them too unreliable for my application as it leads to a complete crash. I'm not sure if it happens more frequently on some operating systems. I will try to investigate this as well as I build for NetBSD, OpenBSD, FreeBSD, Linux, macOS and Windows. At least Linux and Windows have shown stability issues (and I think macOS as well) with these functions but I don't really test so much on BSD Unix so it may happen just as often there. What I know for sure is that it happens at least for 2.0.10 to 2.0.14 as these are what I mostly use.

One of the reasons I doubt the bug is in my code is that I've even had crashes when shutting down my application during debugging sessions, when calling SDL_FreeAudioStream from a destructor. Again, I will try to capture some more data for you but if I recall correctly, the issue was due to a double-free.

To debug with Valgrind is a bit problematic as this is video player code using FFmpeg and it runs amazingly slow in Valgrind and I may have to play hundreds of videos before the crash is triggered.

As for code, I'm not sure I can provide a short snippet, but here is the code that handles audio in my application so you can see what I do:
https://gitlab.com/leonstyhre/emulationstation-de/-/blob/master/es-core/src/AudioManager.cpp

I call these audio functions from my video player:
https://gitlab.com/leonstyhre/emulationstation-de/-/blob/master/es-core/src/components/VideoFFmpegComponent.cpp

**Comment by leonstyhre** [comment]
On the original topic of SDL_AudioStreamAvailable I have no idea how it's implemented, but don't you have the data in some kind of array or similar that you can simply return the size for? In my use case I simply need to know how much data I should ask SDL_AudioStreamGet to retrieve to completely empty the stream.

Another weird thing is that the issue with remaining samples happens quite rarely on Linux and Windows but almost all the time on NetBSD and for some reason it's commonly 880 samples which is a mystery to me.

Anyway, the code snippet I pasted above is completely reliable and has been used for months. Adding 10000 samples always works correctly to return the complete stream size such as 10880. The 10000 is just an arbitrary number by the way, I could just as well have chosen 5000, as long as it's larger than the sample size value :)

**Comment by leonstyhre** [comment]
I managed to reproduce the crash, I simply uncommented the SDL_AudioStreamClear function call and ran my application through a number of videos. After around 50 calls to SDL_AudioStreamClear the crash occured.

It's on Kubuntu 20.04 with this SDL version:
libsdl2-2.0-0:amd64                           2.0.10+dfsg1-3                              amd64        Simple DirectMedia Layer

The error message is:
corrupted size vs. prev_size while consolidating

I honestly have no idea what that means and I'm not experienced with debugging libraries so I'm not sure what data you need from me. But here is at least the call stack from VSCode, maybe it gives some hints:

libc.so.6!__GI_raise(int sig) (/build/glibc-eX1tMB/glibc-2.31/sysdeps/unix/sysv/linux/raise.c:50)
libc.so.6!__GI_abort() (/build/glibc-eX1tMB/glibc-2.31/stdlib/abort.c:79)
libc.so.6!__libc_message(enum __libc_message_action action, const char * fmt) (/build/glibc-eX1tMB/glibc-2.31/sysdeps/posix/libc_fatal.c:155)
libc.so.6!malloc_printerr(const char * str) (/build/glibc-eX1tMB/glibc-2.31/malloc/malloc.c:5347)
libc.so.6!_int_free(mstate av, mchunkptr p, int have_lock) (/build/glibc-eX1tMB/glibc-2.31/malloc/malloc.c:4332)
libSDL2-2.0.so.0![Unknown/Just-In-Time compiled code] (Unknown Source:0)
AudioManager::clearStream(AudioManager * this) (/home/leon/Leon/Programmering/EmulationStation-DE/emulationstation-de/es-core/src/AudioManager.cpp:307)
VideoFFmpegComponent::stopVideo(VideoFFmpegComponent * this) (/home/leon/Leon/Programmering/EmulationStation-DE/emulationstation-de/es-core/src/components/VideoFFmpegComponent.cpp:1407)
VideoFFmpegComponent::~VideoFFmpegComponent(VideoFFmpegComponent * this) (/home/leon/Leon/Programmering/EmulationStation-DE/emulationstation-de/es-core/src/components/VideoFFmpegComponent.cpp:58)
VideoFFmpegComponent::~VideoFFmpegComponent(VideoFFmpegComponent * this) (/home/leon/Leon/Programmering/EmulationStation-DE/emulationstation-de/es-core/src/components/VideoFFmpegComponent.cpp:58)
SystemScreensaver::stopScreensaver(SystemScreensaver * this) (/home/leon/Leon/Programmering/EmulationStation-DE/emulationstation-de/es-app/src/SystemScreensaver.cpp:198)
SystemScreensaver::nextGame(SystemScreensaver * this) (/home/leon/Leon/Programmering/EmulationStation-DE/emulationstation-de/es-app/src/SystemScreensaver.cpp:216)
SystemScreensaver::update(SystemScreensaver * this, int deltaTime) (/home/leon/Leon/Programmering/EmulationStation-DE/emulationstation-de/es-app/src/SystemScreensaver.cpp:423)
Window::update(Window * this, int deltaTime) (/home/leon/Leon/Programmering/EmulationStation-DE/emulationstation-de/es-core/src/Window.cpp:389)
main(int argc, char ** argv) (/home/leon/Leon/Programmering/EmulationStation-DE/emulationstation-de/es-app/src/main.cpp:693)

**Comment by slouken** [comment]
Yep, that's memory corruption. Unfortunately there's no indication of where it happened. You'll need another tool like valgrind or asan to track down where it's coming from.

**Comment by icculus** [comment]
AddressSanitizer should be able to run in real time and hopefully find this issue, but you'll have to build your app with it enabled.

And even without a crash, it's possible Valgrind will point out the bug on the first run, as an invalid access will often still get noticed even if it doesn't crash the whole thing.

**Comment by leonstyhre** [comment]
I'm attaching the valgrind output as I'm not sure what to look for exactly. There are some small memory leaks but I've been unable to fix them and I think they are all outside my application.

There are some warnings likely related to SDL such as a conditional jump based on an uninitialized value. I've looked into that in the past but I think I can't fix it in my code. However I think that's a separate topic not related to the crashes discussed here?

To clarify, this output is from running a couple of videos which includes calls to SDL_AudioStreamClear. The crash did not occur in this valgrind session.

I will try the Clang AddressSanitizer next while triggering the crash.

[valgrind_run_01.txt](https://github.com/libsdl-org/SDL/files/7531719/valgrind_run_01.txt)

**Comment by leonstyhre** [comment]
I managed to capture the crash using AddressSanitizer, please see the attached output.
This is the double-free I recalled seeing some time ago.

It's on Kubuntu, SDL version 2.0.10+dfsg1-3

[asan_output.txt](https://github.com/libsdl-org/SDL/files/7532225/asan_output.txt)

**Comment by sezero** [comment]
If SDL is built with debug symbols, the valgrind and asan outputs will show the proper locations in SDL library.

**Comment by leonstyhre** [comment]
I'm pretty sure I've seen the issue with 2.0.14 on Windows and macOS as well (sorry this was some time ago so I can't recall exactly).

But I'll take a look and see if I can build a debug version from the main branch and capture what actually happens.

**Comment by leonstyhre** [comment]
I think the same bug exists in the main branch, as I think I've reproduced it now. But it shows up as an assertion instead of a crash.
Exactly the same situation though, it appeared after playing around 50 videos, i.e. after around 50 calls to SDL_AudioStreamClear

This is the error:

WARN: 

Assertion failure at SDL_ReadFromDataQueue (/home/leon/Leon/Programmering/reference_software/SDL/src/SDL_dataqueue.c:283), triggered 1 time:
  '(queue->head != ((void *)0)) == (queue->queued_bytes != 0)'

**Comment by slime73** [comment]
Have you confirmed that there isn't a race condition or other issue in your own code that's causing invalid behaviour?

**Comment by leonstyhre** [comment]
I also suspected threading issues in the past so I moved all the calls to the audio routines to the main thread instead of calling them directly from the video decoding thread, but it didn't make any difference as far as the crashes were concerned.

But even if there was a race condition in my code, could it even trigger this issue? SDL_AudioStreamClear should be possible to call even if there is no data in the stream without causing issues as far as I've understood? I mean, it clears the stream of data, so what could possibly a calling application do to cause such an assertion?

**Comment by leonstyhre** [comment]
As well I've had crashes when calling SDL_FreeAudioStream from the destructor in AudioManager on application shutdown. I can't think how that could be caused by my application because that is very late in the shutdown of the application so it's impossible that something attempts to output audio at that time.

**Comment by icculus** [comment]
Just to be clear, is this particular SDL_AudioStream referenced at AudioManager.cpp:307 a global?

It looks like you're calling SDL_FreeAudioStream during a destructor that has destructed a different object previously that had a copy of the same SDL_AudioStream pointer.

**Comment by leonstyhre** [comment]
It's a static variable, but there is only ever a single instance of AudioManager and the only time the destructor is called is during application shutdown. But the assertion above happened while calling SDL_AudioStreamClear, and that's what has caused 99% of the crashes in the past, not the call to SDL_FreeAudioStream (although that has occasionally caused crashes as well).

Just to clarify, when I now use the  SDL_AudioStreamGet method to empty the stream, the code is completely stable and has remained so for 6 months or so, I have not seen a single crash in the video player since and neither valgrind or asan indicates any issues. Not trying to say my code is perfect but it's important for context that everything works fine as long as I don't call SDL_AudioStreamClear, SDL_AudioStreamFlush or SDL_FreeAudioStream :)

**Comment by bradallred** [comment]
It might be useful to do a run with ThreadSanitizer to see what sort of races you are dealing with.

**Comment by bradallred** [comment]
you seem to be attempting to use a `bool` as a callback lock; you should be using  [SDL_LockAudioDevice](https://wiki.libsdl.org/SDL_LockAudioDevice).

**Comment by leonstyhre** [comment]
Thanks for the feedback, I will modify the code as per your recommendations and see if it fixes the problem!

**Comment by icculus** [comment]
Closing this for now. If this turns out to need more SDL work, let us know!
